### PR TITLE
Update CORS middleware to use endpoint metadata

### PR DESCRIPTION
**Pull request by JamesNK**
- Fetch policy name from endpoint metadata
 - Run functional tests with endpoint routing and traditional routing

Addresses https://github.com/aspnet/AspNetCore/issues/4210

### CODE REVIEW

**Comment by JamesNK** [comment]
Design questions:

1. Right now `IDisableCorsAttribute` is not being used. Should it be? If so:

    * Does `IDisableCorsAttribute` override `IEnableCorsAttribute` when both are present? 
    * Or does `IDisableCorsAttribute` override `IEnableCorsAttribute` when it has a higher precedence?

2. The middleware is currently on for every request. This is its previous behavior. `IEnableCorsAttribute` metadata only allows the policy to be customized per endpoint. Should it be changed to be opt in?

**Comment by JamesNK** [comment]
> The middleware is currently on for every request. This is its previous behavior. IEnableCorsAttribute metadata only allows the policy to be customized per endpoint. Should it be changed to be opt in?

Scratch that. While it is on for every request, the default policy name will resolve to `__DefaultCorsPolicy` that has no value in the policy map and the middleware will no-op (apart from a log message saying that it can't find the policy name). The middleware is kind of opt-in today 👍 I think the logging should be changed to when the policy name that can't be found isn't the default policy name.

File: src/CORS/test/Microsoft.AspNetCore.Cors.Test/CorsMiddlewareTests.cs
Comment (dallancarr) [inline src/CORS/test/Microsoft.AspNetCore.Cors.Test/CorsMiddlewareTests.cs]: This test name doesnt sound right 🙄

Comment (dallancarr) [reply src/CORS/test/Microsoft.AspNetCore.Cors.Test/CorsMiddlewareTests.cs]: Ignore me it does 😁

File: src/CORS/samples/SampleDestination/Startup.cs
Comment (rynowak) [inline src/CORS/samples/SampleDestination/Startup.cs]: I'm not `RequireCors` is right. CORS is something you allow, not require 😆 Maybe `AllowCors` or `WithCorsPolicy`.

I think let's just pick something reasonable for now and do API review later. This is startup so it's going to get lots of eyeballs.

File: src/CORS/samples/SampleDestination/Startup.cs
Comment (rynowak) [inline src/CORS/samples/SampleDestination/Startup.cs]: I'm not sure about this pattern. This doesn't work for many many scenarios.

**Comment by rynowak** [comment]
> The middleware is currently on for every request. This is its previous behavior.

What does it do in that case? Does it have a default policy that it applies unless overridden? That seems super nice.

**Comment by rynowak** [comment]
> The middleware is currently on for every request. This is its previous behavior.

What does it do in that case? Does it have a default policy that it applies unless overridden? That seems super nice.

**Comment by rynowak** [comment]
> The middleware is currently on for every request. This is its previous behavior.

What does it do in that case? Does it have a default policy that it applies unless overridden? That seems super nice.

**Comment by rynowak** [comment]
Github is telling me that my comments don't work, but then adding them anyway. Leaving the dupes because it's more A E S T H E T I C

File: src/CORS/src/Microsoft.AspNetCore.Cors/Infrastructure/CorsEndpointConventionBuilderExtensions.cs
Comment (rynowak) [inline src/CORS/src/Microsoft.AspNetCore.Cors/Infrastructure/CorsEndpointConventionBuilderExtensions.cs]: can policy name by null? What does that mean?

File: src/CORS/src/Microsoft.AspNetCore.Cors/Infrastructure/CorsEndpointConventionBuilderExtensions.cs
Comment (rynowak) [inline src/CORS/src/Microsoft.AspNetCore.Cors/Infrastructure/CorsEndpointConventionBuilderExtensions.cs]: Sort order is wrong

File: src/CORS/src/Microsoft.AspNetCore.Cors/Infrastructure/CorsMiddleware.cs
Comment (rynowak) [inline src/CORS/src/Microsoft.AspNetCore.Cors/Infrastructure/CorsMiddleware.cs]: re: `IDisableCorsAttribute` - what I'd imagine is that we need to add a marker interface that both of these implement, and check the type of the most significant one.

I would expect `IDisableCorsAttribute` to work, it's a feature MVC supports

**Review by rynowak — COMMENTED** [review]
This looks on the right track 😁

**Comment by natemcmaster** [comment]
We're adding required PR checks to ensure the repo compiles before you merge. Can you update this PR by rebasing on master?

```
git fetch origin master:master
git rebase master
```

Alternative: if you plan to 'squash merge' this PR, you can also use a merge to update this PR. This can be easier to manage if you have lots of changes in this PR.
```
git fetch origin master:master
git merge master
```

File: src/CORS/src/Microsoft.AspNetCore.Cors/Infrastructure/CorsEndpointConventionBuilderExtensions.cs
Comment (JamesNK) [reply src/CORS/src/Microsoft.AspNetCore.Cors/Infrastructure/CorsEndpointConventionBuilderExtensions.cs]: A null policy name means either the default policy or default policy name configured passed to the middleware constructor is used.

**Comment by JamesNK** [comment]
@natemcmaster The build is failing with this message:

> Unable to find package Microsoft.AspNetCore.Routing with version (>= 3.0.0-preview-18606-34) [D:\a\1\s\src\CORS\CORS.sln]

Doe the order that solutions are compiled need to change now that CORS depends on routing?

**Comment by JamesNK** [comment]
🆙 📅

// @rynowak

File: src/CORS/samples/SampleDestination/Startup.cs
Comment (pranavkm) [inline src/CORS/samples/SampleDestination/Startup.cs]: Nice

Comment (pranavkm) [reply src/CORS/samples/SampleDestination/Startup.cs]: Could we add an overload that also accepts an actual instance of the policy e.g. `WithCorsPolicy(builder => ...)`?

File: src/CORS/src/Microsoft.AspNetCore.Cors/Infrastructure/ICorsAttribute.cs
Comment (pranavkm) [inline src/CORS/src/Microsoft.AspNetCore.Cors/Infrastructure/ICorsAttribute.cs]: `internal interface`?

File: src/CORS/samples/SampleDestination/Startup.cs
Comment (JamesNK) [reply src/CORS/samples/SampleDestination/Startup.cs]: Possible. It would require new metadata. The current `IEnableCorsAttribute` only has a name on it.

File: src/CORS/src/Microsoft.AspNetCore.Cors/Infrastructure/ICorsAttribute.cs
Comment (JamesNK) [reply src/CORS/src/Microsoft.AspNetCore.Cors/Infrastructure/ICorsAttribute.cs]: It needs to be public so if anyone else is searching for CORS metadata they can use the marker interface to find the most significant metadata.

**Comment by natemcmaster** [comment]
> Doe the order that solutions are compiled need to change now that CORS depends on routing?

Yes, CORS builds as a peer of Routing, which is why you're getting errors. https://github.com/aspnet/AspNetCore/blob/94e317f411ce618a56a64d1061f02b9eaa4998a3/build/buildorder.props#L18-L19

You can either bump the build order. Or, wait until I merge https://github.com/aspnet/AspNetCore/pull/4449 and rebase on master.

**Comment by JamesNK** [comment]
@natemcmaster 

Updated. It now builds on the server but requires an update. When I attempt to update the local deps I get this error:

```
C:\Development\Source\AspNetCore [jamesnk/cors-endpoint ≡ +0 ~2 -0 !]> ./run.cmd -Path src/CORS/ -LockFile ./korebuild-lock.txt upgrade deps
Using KoreBuild 3.0.0-build-20181206.3
Executing 'C:\Users\James\.dotnet\x64\dotnet.exe msbuild C:\Users\James\.dotnet\buildtools\korebuild\3.0.0-build-20181206.3\KoreBuild.proj -t:UpgradeDependencies -v:n'
Microsoft (R) Build Engine version 15.9.19+g938f3292a0 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 8/12/2018 3:53:10 PM.
     1>Project "C:\Users\James\.dotnet\buildtools\korebuild\3.0.0-build-20181206.3\KoreBuild.proj" on node 1 (UpgradeDependencies target(s)).
     1>C:\Users\James\.dotnet\buildtools\korebuild\3.0.0-build-20181206.3\modules\KoreBuild.Tasks\module.targets(95,5): error : LineupPackageId was not set. [C:\Users\James\.dotnet\buildtools\korebuild\3.0.0-build-20181206.3\KoreBuild.proj]
     1>Done Building Project "C:\Users\James\.dotnet\buildtools\korebuild\3.0.0-build-20181206.3\KoreBuild.proj" (UpgradeDependencies target(s)) -- FAILED.

Build FAILED.

       "C:\Users\James\.dotnet\buildtools\korebuild\3.0.0-build-20181206.3\KoreBuild.proj" (UpgradeDependencies target) (1) ->
       (UpgradeDependencies target) ->
         C:\Users\James\.dotnet\buildtools\korebuild\3.0.0-build-20181206.3\modules\KoreBuild.Tasks\module.targets(95,5): error : LineupPackageId was not set. [C:\Users\James\.dotnet\buildtools\korebuild\3.0.0-build-20181206.3\KoreBuild.proj]

    0 Warning(s)
    1 Error(s)
```

File: src/CORS/src/Microsoft.AspNetCore.Cors/Infrastructure/CorsEndpointConventionBuilderExtensions.cs
Comment (rynowak) [inline src/CORS/src/Microsoft.AspNetCore.Cors/Infrastructure/CorsEndpointConventionBuilderExtensions.cs]: Docs?

File: src/CORS/src/Microsoft.AspNetCore.Cors/Infrastructure/CorsMiddleware.cs
Comment (rynowak) [inline src/CORS/src/Microsoft.AspNetCore.Cors/Infrastructure/CorsMiddleware.cs]: Sounds right to me!

File: src/CORS/src/Microsoft.AspNetCore.Cors/Infrastructure/ICorsAttribute.cs
Comment (rynowak) [reply src/CORS/src/Microsoft.AspNetCore.Cors/Infrastructure/ICorsAttribute.cs]: `ICorsPolicyMetadata`?

File: src/CORS/src/Microsoft.AspNetCore.Cors/Infrastructure/ICorsAttribute.cs
Comment (rynowak) [inline src/CORS/src/Microsoft.AspNetCore.Cors/Infrastructure/ICorsAttribute.cs]: Is this the same namespace as the other types?

Comment (rynowak) [reply src/CORS/src/Microsoft.AspNetCore.Cors/Infrastructure/ICorsAttribute.cs]: Oh it is. Wierd namespace

File: src/CORS/src/Microsoft.AspNetCore.Cors/Infrastructure/CorsMiddleware.cs
Comment (rynowak) [inline src/CORS/src/Microsoft.AspNetCore.Cors/Infrastructure/CorsMiddleware.cs]: Do we need to have the middleware add a token that says "I ran" so we can no op in the filter?

**Review by rynowak — CHANGES_REQUESTED** [review]
This looks super duper! Only a few minor comments

File: src/CORS/samples/SampleDestination/Program.cs
Comment (pranavkm) [inline src/CORS/samples/SampleDestination/Program.cs]: Could we default to this? It's useful to run this app outside of a test when iterating on the code and being unable to `dotnet run` the app might become quickly annoying

File: src/CORS/test/FunctionalTests/CorsMiddlewareFunctionalTest.cs
Comment (pranavkm) [inline src/CORS/test/FunctionalTests/CorsMiddlewareFunctionalTest.cs]: Could we make these two separate tests? `RunClientTests_WithEndpoint` and `RunClientTests_WithoutEndpoint`? Theories can be annoying to debug.

**Comment by JamesNK** [comment]
Rebasing is not a problem. I'm holding off until Middleware.sln builds in VS.

**Comment by JamesNK** [comment]
Middleware.sln is building for me 👍

Rebased

File: src/Middleware/CORS/src/Infrastructure/ICorsAttribute.cs
Comment (rynowak) [inline src/Middleware/CORS/src/Infrastructure/ICorsAttribute.cs]: What about `ICorsMetadata`? Putting attribute in the name seems silly since it's not an attritube.

**Review by rynowak — APPROVED** [review]

File: src/Middleware/CORS/src/Infrastructure/ICorsAttribute.cs
Comment (JamesNK) [reply src/Middleware/CORS/src/Infrastructure/ICorsAttribute.cs]: Already done locally.

Will look like:

* ICorsMetadata
  * IEnableCorsAttribute
  * IDisableCorsAttribute
  * ICorsPolicyMetadata

ICorsPolicyMetadata would have the policy name and a policy on it to support https://github.com/aspnet/AspNetCore/pull/4460#discussion_r239676450. Constructor would only allow one.

Thoughts? An alternate is ICorsPolicyMetadata only have the CorsPolicy on it and we leave IEnableCorsAttribute as the way to set the policy name.

**Pull request merged**
